src/config.rs

```diff
@@ -280,3 +278,169 @@ mod tests {
+// - /usr/lib64/pkcs11/libsofthsm2.so
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct Pkcs11Config {
+    pub module_directories: Vec<String>,
+    pub allowed_module_paths: Vec<String>,
+}
+
+const CONFIGFILE: &str = "ocicrypt.conf";
+const ENVVARNAME: &str = "OCICRYPT_CONFIG";
+const XDGCONFIGHOME: &str = "XDG_CONFIG_HOME";
+const HOME: &str = "HOME";
+
+impl Pkcs11Config {
+
+    fn get_configuration(&self) -> Result<Pkcs11Config> {
+        match std::env::var(ENVVARNAME) {
+            Ok(filename) => {
+                if filename == "internal" {
+                    // TODO
+                    //return getDefaultCryptoConfigOpts()
+                }
+                if std::path::Path::new(&filename).exists() {
+                    return parse_config_file(filename);
+                }
+                // fall out of match
+            },
+            Err(e) => match e {
+                std::env::VarError::NotPresent => (), // fall out of match
+                std::env::VarError::NotUnicode(s) => return Err(anyhow!("")),
+            },
+        };
+
+        match std::env::var(XDGCONFIGHOME) {
+            Ok(envvar) => {
+                let mut p = std::path::PathBuf::from(envvar);
+                p.push(CONFIGFILE);
+                if std::path::Path::new(&p).exists() {
+                    return parse_config_file(p.to_str().unwrap().to_string());
+                }
+                // fall out of match
+            },
+            Err(e) => match e {
+                std::env::VarError::NotPresent => (), // fall out of match
+                std::env::VarError::NotUnicode(s) => return Err(anyhow!("")),
+            },
+        };
+
+        match std::env::var(HOME) {
+            Ok(envvar) => {
+                let mut p = std::path::PathBuf::from(envvar);
+                p.push(".config");
+                p.push(CONFIGFILE);
+                if std::path::Path::new(&p).exists() {
+                    return parse_config_file(p.to_str().unwrap().to_string());
+                }
+                // fall out of match
+            },
+            Err(e) => match e {
+                std::env::VarError::NotPresent => (), // fall out of match
+                std::env::VarError::NotUnicode(s) => return Err(anyhow!("")),
+            },
+        }
+
+        let mut p = std::path::PathBuf::from("/etc");
+        p.push(CONFIGFILE);
+        if std::path::Path::new(&p).exists() {
+            return parse_config_file(p.to_str().unwrap().to_string());
+        }
+        Err(anyhow!(""))
+    }
+
+    fn get_default_crypto_config_opts(&self) -> Result<Pkcs11Config> {
+
+        let mdyaml = get_default_module_directories_yaml("".to_string())?;
+
+        let config = format!("module-directories:\n\
+                              {} \
+                              allowed-module-paths:\n\
+                              {}", mdyaml, mdyaml);
+        parse_pkcs11_config_file(config.as_bytes())
```

Comment:
may not even need this parse-pkcs11-config-file (which takes bytes) if we change this code (by just hard-coding a default yaml config for example).